### DISCUSSION TITLE
MapProperty on collections seem impossible

### DISCUSSION THREAD

**Discussion by JKetelaar** [riok/mapperly#1030] [Q&A] [answered]
So I'm a bit stuck and I'm staring at the documentation without answers at this point...

So say I have an entity class:
```C#
public class MyClass
{
    public ICollection<MyOtherClass> OtherItems { get; set; }
}

public class MyOtherClass
{
    public string Name { get; set; }
}
```

And a view class:
```C#
public class MyViewClass
{
    public IEnumerable<MyOtherViewClass> ViewItems { get; set; }
}

public class MyOtherViewClass
{
    public string SomeName { get; set; }
}
```

Now I want to map these to each other. Of course the mapping of `OtherItems` to `ViewItems` is fairly straight forward with `MapProperty`:
```C#
[MapProperty(nameof(MyClass.OtherItems), nameof(MyViewClass.ViewItems))]
```

But how do I map `Name` to `SomeName`?
I've already tried the following options:

1. Using `MapProperty`
```C#
[MapProperty("MyClass.OtherItems.Name", "MyViewClass.ViewItems.SomeName")]
```
Which does not seem to work;
```
Error RMG005
```

2. Using a custom method on the collection
```C#
    private static IEnumerable<MyOtherViewClass> ToView(this ICollection<MyClass> values)
    {
        return new List<MyOtherViewClass>();
    }
```    
Which is then not used to map.

3. Using a custom method on the collection
```C#
    private static MyOtherViewClass ToView(this MyClass value)
    {
        return new MyOtherViewClass();
    }
```    
Which is then also not used to map.

I'm out of ideas. Am I overseeing something or is this a bug?

**Comment by TimothyMakkison** [comment, marked as answer]
Hey, sorry to hear your stuck, 2/3 would be the recommended way of handling your issue. Unfortunately the function signatures are invalid.

> 1. Using MapProperty
```C#
[MapProperty("MyClass.OtherItems.Name", "MyViewClass.ViewItems.SomeName")]
```

This won't work for two reasons. 

Firstly the member path is invalid as you have specified the type of the source and taget objects (`MyClass` & `MyViewClass`) therefore mapperly will look for members of the name `MyClass` inside `MyClass`.

Secondly `OtherItems` and `ViewItems` are enumerable so its not possible to map `Name` and `SomeName`, what you've done in `2.` and `3.` is the correct way to handle enumerables.

> 2. Using a custom method on the collection

It looks like you've made a mistake in your example you should use `IEnumerable<MyOtherViewClass> ToView(this ICollection<MyOtherClass> values)` which generates: 
```C#
public static partial global::Riok.Mapperly.Sample.MyViewClass MapTo(global::Riok.Mapperly.Sample.MyClass src)
{
    var target = new global::Riok.Mapperly.Sample.MyViewClass();
    target.ViewItems = ToView(src.OtherItems);
    return target;
}
```

> 3. Using a custom method on the collection

Same issue as above:

```c#
[MapProperty(nameof(MyOtherClass.Name), nameof(MyOtherViewClass.SomeName))]
private static partial MyOtherViewClass MapToOtherViewClass(MyOtherClass src);
```
```C#
public static partial global::Riok.Mapperly.Sample.MyViewClass MapTo(global::Riok.Mapperly.Sample.MyClass src)
{
    var target = new global::Riok.Mapperly.Sample.MyViewClass();
    target.ViewItems = MapToIEnumerable(src.OtherItems);
    return target;
}

private static partial global::Riok.Mapperly.Sample.MyOtherViewClass MapToOtherViewClass(global::Riok.Mapperly.Sample.MyOtherClass src)
{
    var target = new global::Riok.Mapperly.Sample.MyOtherViewClass();
    target.SomeName = src.Name;
    return target;
}

private static global::System.Collections.Generic.IEnumerable<global::Riok.Mapperly.Sample.MyOtherViewClass> MapToIEnumerable(global::System.Collections.Generic.ICollection<global::Riok.Mapperly.Sample.MyOtherClass> source)
{
    var target = new global::Riok.Mapperly.Sample.MyOtherViewClass[source.Count];
    var i = 0;
    foreach (var item in source)
    {
        target[i] = MapToOtherViewClass(item);
        i++;
    }
    return target;
}
```

**Reply by JKetelaar** [reply]
This is incredibly informative, thank you so much. I completely missed the fact you can define mappers which are partial, but implicitly called (private).
